src/docker.rs

```diff
@@ -325,10 +328,143 @@ pub fn run(
         }
     }
 
+    let mut image = image(config, target)?;
+
+    // TODO: if -p, use that
+    let package = metadata.packages.get(0).unwrap();
+
+    if let Some(dockerfile) = config.dockerfile(target)? {
+        image = Dockerfile::File {
+            path: &dockerfile,
+            context: config.dockerfile_context(target)?.as_deref(),
+            name: config.image(target)?.as_deref(),
+        }
+        .build(
+            verbose,
+            &host_root,
+            config.dockerfile_build_args(target)?.unwrap_or_default(),
+            target,
+            package,
+        )
+        .wrap_err("when building dockerfile")?;
+    }
+
+    if let Some(pre_build) = config.pre_build(target)? {
+        let stdin = Dockerfile::Stdin {
+            content: format!(
+                "
+FROM {image}
+ARG CMD
+RUN $CMD"
+            ),
+        };
+        stdin
+            .build(
+                verbose,
+                &host_root,
+                Some(("CMD", pre_build.join("\n"))),
+                target,
+                package,
+            )
+            .wrap_err("when pre-building")?;
+        image = stdin.image_name(target, package);
+    }
+
     docker
-        .arg(&image(config, target)?)
+        .arg(&*image)
         .args(&["sh", "-c", &format!("PATH=$PATH:/rust/bin {:?}", cmd)])
-        .run_and_get_status(verbose)
+        .run_and_get_status(verbose, false)
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Dockerfile<'a> {
+    File {
+        path: &'a str,
+        context: Option<&'a str>,
+        name: Option<&'a str>,
+    },
+    Stdin {
+        content: String,
+    },
+}
+
+impl<'a> Dockerfile<'a> {
+    pub fn build(
+        &self,
+        verbose: bool,
+        host_root: &Path,
+        build_args: impl IntoIterator<Item = (impl AsRef<str>, impl AsRef<str>)>,
+        target_triple: &Target,
+        package: &crate::cargo::Package,
+    ) -> Result<String> {
+        let mut docker_build = docker_command("build")?;
+        docker_build.current_dir(host_root);
+        docker_build.env("DOCKER_SCAN_SUGGEST", "false");
+        let image_name = self.image_name(target_triple, package);
+        docker_build.args(["--tag", &image_name]);
+
+        for (key, arg) in build_args.into_iter() {
+            docker_build.args(["--build-arg", &format!("{}={}", key.as_ref(), arg.as_ref())]);
+        }
+
+        match self {
+            Dockerfile::File { path, .. } => {
```

Comment:
Would it be preferable to handle the `FROM {{image}}` ourselves? That way the users can just add the components they want afterwards? Not sure about this but I think it might be easier.

Reply:
I fail to see the point, you could already do that with 

```toml
pre-build = ["apt-get update && apt-get install --assume-yes --no-install-recommends libclang-3.9-dev clang-3.9"]
```

I see another solution though, make a `target.<target>.dockerfile.provide-base` that would do it instead with the image you provide. But I still don't see the benefit

Reply:
That's just a simple example: if something requires the `ENV` keyword or similar (like, say we want `vcpkg` for providing dependencies of cross-compiled C/C++ libraries), we can't simply use pre-build hooks.

Reply:
A more complex example would be:

```Dockerfile
RUN apt-get update && apt-get install --assume-yes --no-install-recommends \
    libclang-3.9-dev clang-3.9
RUN git clone https://github.com/Microsoft/vcpkg.git && ./vcpkg/bootstrap-vcpkg.sh
ENV PATH /path/to/vcpkg/bin/dir/:$PATH

# do more stuff here, maybe add an ARG, etc.
# everything is platform generic.
```

Reply:
Maybe we could have `pre-build` hooks encompass this functionality (`RUN` isn't implicit anymore, and then all this can be easily encompassed there)? That way you can provide a complete Dockerfile, or just add extra commands to the existing one?

Reply:
I've implemented this (i think, not fully tested), but with allowing the user to tell cross to insert `FROM` with `target.TARGET.dockerfile.provide-base`.

Reply:
and removed with f98a6e630507ebfd92af5dbbfbd72e449dc72e7c

instead we provide a build arg

`ARG CROSS_BUILD_IMAGE` that will be whatever cross would use